### PR TITLE
Add semantic reranking functionality to CognitiveSearchVectorStore

### PR DESCRIPTION
**Pull request by noppadol-assava**
# Description

This pull request is to add a semantic reranking option to CognitiveSearchVectorStore.  
Please refer to this https://learn.microsoft.com/en-us/azure/search/search-get-started-semantic?tabs=python for more detail on semantic reranking.
The `query()` method is also refactored to better support multiple search types.

Example Usage
`index.as_retriever(vector_store_query_mode=VectorStoreQueryMode.SEMANTIC_HYBRID)`

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/vector_stores/cogsearch.py [hunk 1]
```diff
@@ -515,49 +515,183 @@ def _create_odata_filter(self, metadata_filters: MetadataFilters) -> str:
...
+        )
+        if query.mode == VectorStoreQueryMode.SPARSE:
+            azure_query_result_search = AzureQueryResultSearchSparse(
+                query, self._field_mapping, odata_filter, self._search_client
+            )
+        elif query.mode == VectorStoreQueryMode.HYBRID:
+            azure_query_result_search = AzureQueryResultSearchHybrid(
+                query, self._field_mapping, odata_filter, self._search_client
+            )
+        elif query.mode == VectorStoreQueryMode.SEMANTIC_HYBRID:
```
Comment (hatianzhang) [inline llama_index/vector_stores/cogsearch.py:534]: is there any default query mode?

**Review by hatianzhang — COMMENTED** [review]
this is awesome. could you also update `/llama_index/docs/examples/vector_stores/CognitiveSearchIndexDemo.ipynb` to showcase those different search mode and reranking

**Comment by noppadol-assava** [comment]
> this is awesome. could you also update `/llama_index/docs/examples/vector_stores/CognitiveSearchIndexDemo.ipynb` to showcase those different search mode and reranking

Added please review.

**Review by hatianzhang — APPROVED** [review]
LGTM!

**Pull request merged**